### PR TITLE
Adds Utility Obj

### PR DESCRIPTION
**Pull request by mlzummo**
@GordonSmith @jbrundage @dtsnell4 

Please Review

Signed-off-by: Mathew Zummo <mzummo@gmail.com>

### CODE REVIEW

**Comment by GordonSmith** [comment]
Would common be a better location for this?

**Comment by jbrundage** [comment]
I also would have assumed it should go in "other"

**Comment by GordonSmith** [comment]
Currently "other" depends on "common" - If we ever use a Util function in common, then that dependency will be circular (which is technically ok - but flags to me that Util may belong in common).

**Comment by mlzummo** [comment]
@GordonSmith Moved to common/

File: src/common/Utility.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+                oFxNcL += '';
+                oFyNcL += '';
+            }
+            if (oFxNcL < oFyNcL) { return order === "ascending" ? -1 : 1; }
+            if (oFxNcL > oFyNcL) { return order === "ascending" ? 1 : -1; }
+        }
+        return 0;
+    };
+
+    return Utility;
```
Comment (GordonSmith) [inline src/common/Utility.js]: There should be no "Utility" function and then your content will simply be:
```
function _naturalSort(...) {
  //  This function is not accessable outside of this file.
}
return {
    naturalSort: function(...) {
        ...
    }
};
```

File: src/common/Utility.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
+"use strict";
+(function (root, factory) {
+    if (typeof define === "function" && define.amd) {
+        define([], factory);
+    } else {
+        root.common_Utility = factory();
+    }
+}(this, function () {
+    function Utility() {
```
Comment (GordonSmith) [inline src/common/Utility.js]: Remove

File: src/common/Utility.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    /**
+     * USAGE: this._data = Utility.sort(this._data, "naturalsort", context._currentSortOrder === 1 ? "ascending" : "descending", idx, true);
+     */
+    Utility.sort = function (data, type, order, idx, sortCaseSensitive, options) {
```
Comment (GordonSmith) [inline src/common/Utility.js]: rename to naturalSort

File: src/common/Utility.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        var context = this;
+        if (!type) { type = "natualsort"; }
+        return data.sort(function(a,b) {
+            switch (type) {
```
Comment (GordonSmith) [inline src/common/Utility.js]: remove switch statement that doesn't do anything

**Comment by GordonSmith** [comment]
Looking at the code I would suggest refactoring it and expose it as a "comparitor" like d3.ascending - then you can re-use it to sort and Array of Strings as well as an Array of Objects,

**Comment by mlzummo** [comment]
how will I be able to handle case sensitivity and later adding the option of how to sort numbers (beginning or end)

**Comment by GordonSmith** [comment]
Same as you are now?  I am not suggesting deleting the code, just refactoring.  As it stands I can't use it to sort the following:  ```["A100", "A10", "A1"]```

File: src/common/Utility.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    if (typeof define === "function" && define.amd) {
+        define([], factory);
+    } else {
+        root.common_Utility = factory();
+    }
+}(this, function () {
+
+    /**
+     *
+     */
```
Comment (GordonSmith) [inline src/common/Utility.js]: Remove empty comments

File: src/common/Utility.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+            }
+            if (oFxNcL < oFyNcL) { return order === "ascending" ? -1 : 1; }
+            if (oFxNcL > oFyNcL) { return order === "ascending" ? 1 : -1; }
+        }
+        return 0;
+    }
+
+    return {
+        /*
+         * USAGE: this._data = Utility.sort(this._data, context._currentSortOrder === 1 ? "ascending" : "descending", idx, true);
```
Comment (GordonSmith) [inline src/common/Utility.js]: I am not sure how I feel about this comment - it does have some useful information (ascending and descending are two options for "order") but the user still needs to dig into the code to workout how to use the function (which is fine IMO).

File: src/common/Utility.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+         */
+        naturalSort: function(data, order, idx, sortCaseSensitive, options) {
```
Comment (GordonSmith) [inline src/common/Utility.js]: options not used.

**Comment by GordonSmith** [comment]
Some thoughts on inline JSDoc comments:
* They need to add value.
* They need to be 100% accurate and 100% complete  - the developer needs to be confident in this otherwise they will still just dig into the sources to see what it does - at which point the comment becomes noise.
* In my experience comments over time get out of sync with the sources.
* A wrong comment can be just as damaging as a bug (more so in some respects) and are impossible to test.

**Comment by GordonSmith** [comment]
Can you add some unit tests for this (the framework for this just got added over the weekend, so ping me I can show you where/how).

Further:  A set of unitTests also doubles as "documentation" as it will clearly show working examples and is guaranteed to be in sync with the sources...

File: test/runner.html [hunk 8]
```diff
@@ -91,6 +91,9 @@
         require(["./widgets"], function () {
             mocha.run();
         });
+        require(["./utility"], function () {
+            mocha.run();
+        });
     </script>
```
Comment (GordonSmith) [inline test/runner.html]: You should combine those to:  
```
require(["./widgets", "./utility"], function () {
           mocha.run();
        });
```

File: test/utility.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+            var sortedListD = Utility.naturalSort(list,"descending",null,true);
+            var expectedListD = [456, "222", 123, "111", "c", "b", "a", "CD456", "AB123", "AB", "AA"];
+
+            assert.deepEqual(sortedListA,expectedListA);
+            assert.deepEqual(sortedListB,expectedListB);
+            assert.deepEqual(sortedListC,expectedListC);
+            assert.deepEqual(sortedListD,expectedListD);
+
+            done();
```
Comment (GordonSmith) [inline test/utility.js]: You haven't tested the 3rd param?

File: test/runner.html (hunk 8 above)
Comment (mlzummo) [inline test/runner.html]: is there a quick way we can turn on an off a group then? the way it is now i can quickly comment out the moch.run() for any particular one

Comment (mlzummo) [inline test/runner.html]: http://localhost:8080/test/runner.html?grep=utility    ?

**Comment by GordonSmith** [comment]
@mlzummo Can you rebase.

**Pull request merged**